### PR TITLE
More user factory replacements

### PR DESCRIPTION
**Pull request by gravitystorm**
Slowly getting there, but still more to come...

### CODE REVIEW

File: test/controllers/trace_controller_test.rb [hunk 1]
```diff
@@ -389,115 +394,115 @@ def test_data_not_found
...
   # Test downloading the picture for a trace that doesn't exist
   def test_picture_not_found
     # First with no auth, which should work since the trace is public
-    get :picture, :display_name => users(:public_user).display_name, :id => 0
+    get :picture, :display_name => create(:user).display_name, :id => 0
     assert_response :not_found
 
     # Now with some other user, which should work since the trace is public
-    get :picture, { :display_name => users(:public_user).display_name, :id => 0 }, { :user => users(:public_user) }
+    get :picture, { :display_name => create(:user).display_name, :id => 0 }, { :user => create(:user) }
```
Comment (tomhughes) [inline test/controllers/trace_controller_test.rb]: That's not doing the same thing any more as it is now authenticating as a different user to the one that nominally owns the non-existent trace.

In the previous test that did the same thing you created the deleted trace up front and used that user throughout.

File: test/controllers/trace_controller_test.rb [hunk 2]
```diff
@@ -389,115 +394,115 @@ def test_data_not_found
...
   # Test downloading the icon for a trace that doesn't exist
   def test_icon_not_found
     # First with no auth
-    get :icon, :display_name => users(:public_user).display_name, :id => 0
+    get :icon, :display_name => create(:user).display_name, :id => 0
     assert_response :not_found
 
     # Now with some other user
-    get :icon, { :display_name => users(:public_user).display_name, :id => 0 }, { :user => users(:public_user) }
+    get :icon, { :display_name => create(:user).display_name, :id => 0 }, { :user => create(:user) }
```
Comment (tomhughes) [inline test/controllers/trace_controller_test.rb]: Same issue here as above.

File: test/controllers/trace_controller_test.rb [hunk 3]
```diff
@@ -554,90 +563,90 @@ def test_create_post
...
+    assert_redirected_to :controller => :user, :action => :login, :referer => trace_edit_path(:display_name => public_trace_file.user.display_name, :id => public_trace_file.id)
 
     # Now with some other user, which should fail
-    get :edit, { :display_name => users(:normal_user).display_name, :id => public_trace_file.id }, { :user => users(:public_user) }
+    get :edit, { :display_name => public_trace_file.user.display_name, :id => public_trace_file.id }, { :user => create(:user) }
     assert_response :forbidden
 
     # Now with a trace which doesn't exist
-    get :edit, { :display_name => users(:public_user).display_name, :id => 0 }, { :user => users(:public_user) }
+    get :edit, { :display_name => create(:user).display_name, :id => 0 }, { :user => create(:user) }
```
Comment (tomhughes) [inline test/controllers/trace_controller_test.rb]: Same again...

File: test/controllers/trace_controller_test.rb [hunk 4]
```diff
@@ -554,90 +563,90 @@ def test_create_post
...
     assert_response :forbidden
 
     # Now with some other user, which should fail
-    post :edit, { :display_name => users(:normal_user).display_name, :id => public_trace_file.id }, { :user => users(:public_user) }
+    post :edit, { :display_name => public_trace_file.user.display_name, :id => public_trace_file.id }, { :user => create(:user) }
     assert_response :forbidden
 
     # Now with a trace which doesn't exist
-    post :edit, { :display_name => users(:public_user).display_name, :id => 0 }, { :user => users(:public_user) }
+    post :edit, { :display_name => create(:user).display_name, :id => 0 }, { :user => create(:user) }
```
Comment (tomhughes) [inline test/controllers/trace_controller_test.rb]: Guess what ;-)

File: test/controllers/trace_controller_test.rb [hunk 5]
```diff
@@ -554,90 +563,90 @@ def test_create_post
...
     assert_response :forbidden
 
     # Now with some other user, which should fail
-    post :edit, { :display_name => users(:normal_user).display_name, :id => public_trace_file.id, :trace => new_details }, { :user => users(:public_user) }
+    post :edit, { :display_name => public_trace_file.user.display_name, :id => public_trace_file.id, :trace => new_details }, { :user => create(:user) }
     assert_response :forbidden
 
     # Now with a trace which doesn't exist
-    post :edit, { :display_name => users(:public_user).display_name, :id => 0 }, { :user => users(:public_user), :trace => new_details }
+    post :edit, { :display_name => create(:user).display_name, :id => 0 }, { :user => create(:user), :trace => new_details }
```
Comment (tomhughes) [inline test/controllers/trace_controller_test.rb]: ♻️

File: test/controllers/trace_controller_test.rb [hunk 6]
```diff
@@ -646,163 +655,163 @@ def test_edit_post_with_details
...
     assert_response :forbidden
 
     # Now with some other user, which should fail
-    post :delete, { :display_name => users(:normal_user).display_name, :id => public_trace_file.id }, { :user => users(:public_user) }
+    post :delete, { :display_name => public_trace_file.user.display_name, :id => public_trace_file.id }, { :user => create(:user) }
     assert_response :forbidden
 
     # Now with a trace which doesn't exist
-    post :delete, { :display_name => users(:public_user).display_name, :id => 0 }, { :user => users(:public_user) }
+    post :delete, { :display_name => create(:user).display_name, :id => 0 }, { :user => create(:user) }
```
Comment (tomhughes) [inline test/controllers/trace_controller_test.rb]: ♻️

File: test/controllers/trace_controller_test.rb [hunk 7]
```diff
@@ -646,163 +655,163 @@ def test_edit_post_with_details
...
     # First with no auth
-    get :api_data, :display_name => users(:public_user).display_name, :id => 0
+    get :api_data, :display_name => create(:user).display_name, :id => 0
     assert_response :unauthorized
 
     # Now with a trace that has never existed
-    basic_authorization(users(:public_user).display_name, "test")
-    get :api_data, :display_name => users(:public_user).display_name, :id => 0
+    basic_authorization(create(:user).display_name, "test")
+    get :api_data, :display_name => create(:user).display_name, :id => 0
```
Comment (tomhughes) [inline test/controllers/trace_controller_test.rb]: ♻️

**Comment by gravitystorm** [comment]
Given the trace doesn't exist, isn't it irrelevant which username is used in the request?

**Comment by tomhughes** [comment]
Well it's true that those whole tests are of dubious validity since the username used in the URL doesn't actually mean anything as you say.

That's a "white box testing" argument of course, where you as tester know what is really going on, while the "black box" argument would be that you don't know that and you're testing whether it behaves differently when logged in as the alleged user. Of course in that case you can then argue for a third test to check what happens when logged in as a different user ;-)

What is definitely true is that the test is not doing the same thing as before, and different tests with essentially the same logic have been converted in inconsistent ways.

**Comment by gravitystorm** [comment]
I realise the display_name is ignored in all cases, but I've been trying to keep the tests reasonably sensible i.e. using the display_name of the owner of the trace. The only tricky bit is dealing with these tests that refer to non-existent traces, since they have no owner.

I've just pushed a change to make them all consistent, i.e. using a fresh create(:user) as the display_name whenever we're referring to a non-existent trace.

**Comment by tomhughes** [comment]
Well if you're going to do that I think you might as well get rid of the second test in each one altogether as it serves no purpose.

**Comment by gravitystorm** [comment]
Done

**Pull request merged**